# Explore RNA methylation using MeRIP-seq
---------------------------------

## **Contents**

+ [Overview](#overview)
+ [Background](#background)
+ [Before Starting](#before-starting)
+ [Getting Started](#getting-started)
+ [Software Requirements](#software-requirements)
+ [Architecture Design](#architecture-design)
+ [Data](#data)
+ [Funding](#funding)
+ [License for Data](#license-for-data)

## **Overview**


## **Background**
The epitranscriptome encompasses all biochemical modifications of RNA in a cell, representing a rapidly evolving area of study within molecular biology. While the discovery of RNA modifications dates back to the 1950s, initial research predominantly focused on highly abundant RNA species like ribosomal RNA (rRNA) and transfer RNA (tRNA). These molecules were easier to study due to their abundance and structural simplicity. However, low-abundance RNAs, such as messenger RNA (mRNA), remained largely unexplored due to technological limitations in detecting modifications on these less common species.

The landscape of RNA modifications shifted dramatically in 2011 when scientists discovered that N6-methyladenosine (m6A), one of the most common RNA modifications, is reversible. This discovery, driven by the identification of the FTO protein as an m6A demethylase, revealed the dynamic nature of RNA modifications. It suggested that m6A could be actively added and removed in response to cellular needs, much like modifications in the epigenome. This breakthrough revitalized interest in the field of epitranscriptomics and highlighted RNA methylation as a key regulatory layer in gene expression.

In 2012, the development of high-throughput profiling techniques, including MeRIP-seq (Methylated RNA Immunoprecipitation Sequencing), allowed for the first transcriptome-wide mapping of m6A. These approaches, which utilize m6A-specific antibodies, uncovered the widespread presence of m6A on various RNA species and demonstrated its role in dynamically responding to cellular conditions. The ability to detect m6A modifications across the entire transcriptome provided unprecedented insight into the regulatory roles of m6A in processes such as RNA stability, translation, and splicing.

Even today, antibody-based m6A detection remains a predominant and straightforward method for studying RNA methylation, continuing to drive advancements in understanding how RNA modifications contribute to cellular functions and diseases.

**RNA methylation** is a form of post-transcriptional modification that occurs on RNA molecules. The most common and well-studied type of RNA methylation is N6-methyladenosine (m6A), where a methyl group (-CH3) is added to the nitrogen at position 6 of adenosine bases in RNA. RNA methylation plays critical roles in various biological processes, such as: mRNA stability, translation regulation, RNA splicing and RNA localization. This modification is dynamically regulated by writers (methyltransferases that add m6A), erasers (demethylases that remove m6A), and readers (proteins that recognize and bind to m6A to mediate its functions). 
<img src="images/RNA-methylation.png" width="600" />

**MeRIP-seq** (Methylated RNA Immunoprecipitation Sequencing) is a high-throughput technique used to map RNA methylation, particularly N6-methyladenosine (m6A), across the transcriptome. This method combines immunoprecipitation, using an m6A-specific antibody, with RNA sequencing to identify RNA fragments that are enriched for m6A modifications. The process begins by pulling down m6A-marked RNA fragments from total RNA using the antibody, followed by sequencing both the immunoprecipitated (IP) and input RNA libraries. By comparing these datasets, researchers can pinpoint the locations of m6A modifications, referred to as m6A peaks, along RNA transcripts. MeRIP-seq provides a powerful tool for studying the role of m6A in regulating RNA stability, translation, splicing, and other post-transcriptional processes, offering new insights into gene expression regulation in development, disease, and response to environmental changes.

<img src="images/MeRIP-seq.png" width="400" />

## **Before Starting**


## **Getting Started**


## **Software Requirements**


## **Architecture Design**


## **Data**


## **Funding**


## **License for Data**

