{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <b>Module 4. nf-meripseq - an integrated nexflow pipeline </b>\n",
    "--------------------------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Overview\n",
    "<b>nf-meripseq</b> is an integrated analysis pipeline for MeRIP-seq data built on <a href=\"https://www.nextflow.io/docs/latest/index.html\">Nextflow</a>, with full support from the <a href =\"https://nf-co.re/\">nf-core</a> community. It provides an automated, user-friendly solution for in-depth mining of MeRIP-seq data, focusing on m6A modification analysis. The pipeline integrates a wide range of functional modules, from basic processing (e.g., alignment, peak calling) to advanced downstream analysis (e.g., differential expression, visualization). As an nf-core-supported pipeline, it adheres to best-practice standards and leverages Nextflow with <a href=\"https://docs.docker.com/\">Docker</a> and <a href=\"https://docs.sylabs.io/guides/4.3/user-guide/\">Singularity</a>  support, ensuring high reproducibility and scalability across diverse computing environments. nf-meripseq is ideal for processing large sample sets with a single command, delivering results in a structured output directory organized by step and tool. Comprehensive visualization reports, including tables and plots, are provided as HTML files, benefiting from nf-coreâ€™s rigorous testing and documentation framework."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Learning Objectives\n",
    "+ Get familiar with nextflow and nf-core workflows\n",
    "+ Run nf-meripseq using the example dataset\n",
    "    - Understanding the input paramerters of the nf-meripseq workflow\n",
    "    - Understanding the output and reports from the nf-meripseq workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prerequisites\n",
    "- Understanding the basic processing steps of MeRIPseq data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get Started\n",
    "All the analysis modules are generated by Nextflow, and all the third-party tools are encapsulated in the Docker container.\n",
    "\n",
    "### 1. Install necessary package \n",
    "install <code>nextflow</code>, <code>Singularity</code> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! conda install bioconda::nextflow conda-forge::singularity -y --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Get datasets\n",
    "This example dataset is the same one used in submodule1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# copy the data from s3 bucket to example_dataset directory\n",
    "! aws s3 cp s3://nigms-sandbox/ovarian-cancer-example-fastqs/ example_dataset --recursive\n",
    "# decompress the sequence reads files\n",
    "! tar -zxvf example_dataset/fastqs.tar.gz -C example_dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Run nf-meripseq\n",
    "<b>samplesheet.csv</b> \n",
    "The <code>control</code> column should be the sample identifier for the controls for any given IP. This column together with the control_replicate column will set the corresponding control for each of the samples in the table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! nextflow run nf-meripseq -profile singularity \\\n",
    "    --input example_dataset/samplesheet.csv \\\n",
    "    --gtf example_dataset/gencode.v46.pri.chr11.1.5M.gtf \\\n",
    "    --fasta example_dataset/chr11_1.5M.fasta \\\n",
    "    --read_length 37 \\\n",
    "    --outdir=\"meripseq_results\" \\\n",
    "    -c add.config \\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusion\n",
    "Provide an overview of the lessons and skills learned from the module."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean up\n",
    "A reminder to shutdown VM and delete any relevant resources. <br><br>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_python3",
   "language": "python",
   "name": "conda_python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
